runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java

```diff
@@ -259,19 +270,23 @@ private void detectTranslationMode(Pipeline pipeline) {
   }
 
   /**
-   * Traverses the Pipeline to determine the {@link TranslationMode} for this pipeline.
+   * Traverses the Pipeline to determine the {@link TranslationMode} for this pipeline,
+   * and populate the candidates for caching. It's the preparation step of the runner.
    */
-  static class TranslationModeDetector extends Pipeline.PipelineVisitor.Defaults {
-    private static final Logger LOG = LoggerFactory.getLogger(TranslationModeDetector.class);
+  @VisibleForTesting
+  static class DAGPreVisit extends Pipeline.PipelineVisitor.Defaults {
+    private static final Logger LOG = LoggerFactory.getLogger(DAGPreVisit.class);
 
     private TranslationMode translationMode;
+    private Map<PCollection, Long> cacheCandidates;
 
-    TranslationModeDetector(TranslationMode defaultMode) {
+    DAGPreVisit(TranslationMode defaultMode, Map<PCollection, Long> cacheCandidates) {
```

Comment:
Got your point. Not problem to have `cacheCandidate`  as member of `DAGPreVisit` instead of `SparkRunner`.